File: Web/packages/kernel/test/livelogs-conformance.test.ts
//
//  livelogs-conformance.test.ts — the SHARED live-logs corpus through the TS kernel
//  (OpenSource/Conformance/livelogs/{wire,report}.json). The Kotlin twin (:core
//  LiveLogsConformanceTest) and the Swift twin (LiveLogs) run the SAME files, so what leaves a
//  device, how the relay assigns and replays seq, and what the report verifier says about a
//  pasted blob cannot differ by renderer — the verdicts ARE the support macro, and a verdict
//  that flips per platform would re-open the exact incident the feature exists to close.
//
//  Missing corpus = loud failure — a silently-skipped conformance suite is how drift starts.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  LIVE_WIRE_VERSION, LIVE_MESSAGE_CAP, LIVE_BATCH_MAX_ROWS, LIVE_QUEUE_CAP,
  LIVE_IDLE_ACK_PAUSE, LIVE_RING_CAP,
  liveRowFromLog, liveRowFromError, liveRowFromKernel, liveBatchBody,
  liveAckStart, liveAckFold, liveAckExpire, LiveQueue, LiveRing,
  liveCanonical, liveSha256Hex, liveReportSeal, liveReportVerdict,
  type LiveAckState,
} from "../src/livelogs.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/livelogs");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/livelogs not found");
    dir = parent;
  }
}

function corpus(file: string): { [k: string]: unknown } {
  const doc = JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as { [k: string]: unknown };
  assert.equal(doc["version"], 1, `${file}: version`);
  return doc;
}

function cases<T>(file: string, key: string): T[] {
  const list = corpus(file)[key] as T[];
  assert.ok(Array.isArray(list) && list.length > 0, `${file} ${key} corpus must not be empty`);
  return list;
}

test("livelogs: the constants agree with the corpus", () => {
  const constants = corpus("wire.json")["constants"] as { [k: string]: number };
  assert.equal(constants["wireVersion"], LIVE_WIRE_VERSION);
  assert.equal(constants["messageCap"], LIVE_MESSAGE_CAP);
  assert.equal(constants["batchMaxRows"], LIVE_BATCH_MAX_ROWS);
  assert.equal(constants["queueCap"], LIVE_QUEUE_CAP);
  assert.equal(constants["idleAckPause"], LIVE_IDLE_ACK_PAUSE);
  assert.equal(constants["ringCap"], LIVE_RING_CAP);
});

type RowCase = { name: string; kind: string; entry: { [k: string]: unknown }; at: number; expect: { [k: string]: unknown } };

test("livelogs: every row fold agrees with the corpus", () => {
  for (const testCase of cases<RowCase>("wire.json", "rows")) {
    let row: unknown;
    if (testCase.kind === "log") {
      row = liveRowFromLog(testCase.entry as { scheme: string; level: string; message: string }, testCase.at);
    } else if (testCase.kind === "error") {
      row = liveRowFromError(testCase.entry as {
        scheme: string; code: string; message?: string | null; recoverable: boolean; origin: string;
      }, testCase.at);
    } else if (testCase.kind === "kernel") {
      row = liveRowFromKernel((testCase.entry as { line: string }).line, testCase.at);
    } else {
      throw new Error(`${testCase.name}: unknown kind ${testCase.kind}`);
    }
    assert.deepEqual(row, testCase.expect, testCase.name);
  }
});

type BatchCase = { name: string; sid: string; n: number; rows: never[]; expect?: unknown; canonical?: string };

test("livelogs: the batch body agrees with the corpus", () => {
  for (const testCase of cases<BatchCase>("wire.json", "batch")) {
    const body = liveBatchBody(testCase.sid, testCase.n, testCase.rows);
    if (testCase.expect !== undefined) assert.deepEqual(body, testCase.expect, testCase.name);
    if (testCase.canonical !== undefined) assert.equal(liveCanonical(body), testCase.canonical, testCase.name);
  }
});

type AckStep = { ack?: { ok: boolean; viewers: number; ttlMs: number }; expire?: boolean; at: number };
type AckExpect = { [k: string]: unknown };
type AckCase = { name: string; steps: AckStep[]; expect: AckExpect[] };

test("livelogs: the ack fold agrees with the corpus", () => {
  for (const testCase of cases<AckCase>("wire.json", "ack")) {
    assert.equal(testCase.steps.length, testCase.expect.length, `${testCase.name}: one expectation per step`);
    let state: LiveAckState = liveAckStart();
    testCase.steps.forEach((step, index) => {
      const where = `${testCase.name}: step ${index}`;
      if (step.ack !== undefined) state = liveAckFold(state, step.ack, step.at);
      else if (step.expire === true) state = liveAckExpire(state, step.at);
      else throw new Error(`${where} names no operation`);
      for (const [key, value] of Object.entries(testCase.expect[index]!)) {
        assert.deepEqual((state as unknown as { [k: string]: unknown })[key], value, `${where}: ${key}`);
      }
    });
  }
});

type QueueStep = { push?: string; batch?: number; ack?: number };
type QueueExpect = { size?: number; dropped?: number; batch?: string[] };
type QueueCase = { name: string; cap: number; steps: QueueStep[]; expect: QueueExpect[] };

test("livelogs: the bounded device queue agrees with the corpus", () => {
  for (const testCase of cases<QueueCase>("wire.json", "queue")) {
    assert.equal(testCase.steps.length, testCase.expect.length, `${testCase.name}: one expectation per step`);
    const queue = new LiveQueue<string>(testCase.cap);
    testCase.steps.forEach((step, index) => {
      const expected = testCase.expect[index]!;
      const where = `${testCase.name}: step ${index}`;
      if (step.push !== undefined) {
        const actual = queue.push(step.push);
        assert.equal(actual.size, expected.size, `${where}: size`);
        assert.equal(actual.dropped, expected.dropped, `${where}: dropped`);
      } else if (step.batch !== undefined) {
        assert.deepEqual(queue.batch(step.batch), expected.batch, `${where}: batch`);
      } else if (step.ack !== undefined) {
        assert.equal(queue.ack(step.ack).size, expected.size, `${where}: size after ack`);
      } else {
        throw new Error(`${where} names no operation`);
      }
    });
  }
});

type RingStep = { appendBatch?: { n: number; rows: string[] }; read?: { after: number; limit: number } };
type RingExpect = { accepted?: boolean; last?: number; seqs?: number[]; gap?: boolean };
type RingCase = { name: string; cap: number; steps: RingStep[]; expect: RingExpect[] };

test("livelogs: the relay ring agrees with the corpus", () => {
  for (const testCase of cases<RingCase>("wire.json", "ring")) {
    assert.equal(testCase.steps.length, testCase.expect.length, `${testCase.name}: one expectation per step`);
    const ring = new LiveRing<string>(testCase.cap);
    testCase.steps.forEach((step, index) => {
      const expected = testCase.expect[index]!;
      const where = `${testCase.name}: step ${index}`;
      if (step.appendBatch !== undefined) {
        const actual = ring.appendBatch(step.appendBatch.n, step.appendBatch.rows);
        assert.equal(actual.accepted, expected.accepted, `${where}: accepted`);
        assert.equal(actual.last, expected.last, `${where}: last`);
      } else if (step.read !== undefined) {
        const actual = ring.read(step.read.after, step.read.limit);
        assert.deepEqual(actual.rows.map((entry) => entry.seq), expected.seqs, `${where}: seqs`);
        assert.equal(actual.gap, expected.gap, `${where}: gap`);
      } else {
        throw new Error(`${where} names no operation`);
      }
    });
  }
});

type ShaCase = { input: string; expect: string };

test("livelogs: sha256 agrees with the standard vectors", () => {
  for (const testCase of cases<ShaCase>("report.json", "sha256")) {
    assert.equal(liveSha256Hex(testCase.input), testCase.expect, JSON.stringify(testCase.input.slice(0, 24)));
  }
});

type CanonicalCase = { name: string; value: unknown; expect: string };

test("livelogs: canonical bytes agree with the corpus", () => {
  for (const testCase of cases<CanonicalCase>("report.json", "canonical")) {
    assert.equal(liveCanonical(testCase.value), testCase.expect, testCase.name);
  }
});

test("livelogs: what has no canonical form is refused", () => {
  for (const testCase of cases<{ name: string; value: unknown }>("report.json", "canonicalRejects")) {
    assert.throws(() => liveCanonical(testCase.value), testCase.name);
  }
});

type SealCase = { name: string; body: { [k: string]: unknown }; hash: string; text: string };

test("livelogs: the report seal agrees with the corpus", () => {
  for (const testCase of cases<SealCase>("report.json", "seal")) {
    const sealed = liveReportSeal(testCase.body);
    assert.equal(sealed.hash, testCase.hash, `${testCase.name}: hash`);
    assert.equal(sealed.text, testCase.text, `${testCase.name}: text`);
    assert.deepEqual(liveReportVerdict(sealed.text), { verdict: "genuine", assertion: false },
                     `${testCase.name}: a fresh seal verifies`);
  }
});

type VerdictCase = { name: string; text: string; expect: { verdict: string; assertion: boolean } };

test("livelogs: the verifier verdicts agree with the corpus", () => {
  for (const testCase of cases<VerdictCase>("report.json", "verdict")) {
    assert.deepEqual(liveReportVerdict(testCase.text), testCase.expect, testCase.name);
  }
});

File: Engine/Android/core/src/test/kotlin/despia/engine/LiveLogsConformanceTest.kt
package despia.engine

import java.io.File
import kotlin.math.floor
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * The live-logs conformance runner - executes OpenSource/Conformance/livelogs/{wire,report}.json
 * through THIS runtime's LiveLogs / AckState / LiveQueue / LiveRing (proposals/live-logs.md). The
 * TS twin (@despia/kernel livelogs.ts, livelogs-conformance.test.ts) and the Swift reference
 * (LiveLogs.swift) run the SAME files, so what leaves a device, how the relay paces it, and what
 * a `.dsxreport` seal verifies as cannot drift between renderers - a scrubber that fires on one
 * platform and not another is a privacy incident with a platform column, and two ends of one wire
 * that disagree about an idempotency key double every retried row.
 *
 * Missing corpus = loud failure - a silently-skipped conformance suite is how drift starts.
 */
class LiveLogsConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/livelogs/$name")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/livelogs/$name not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(name: String): Map<String, Any?> {
        val root = json(corpusFile(name).readText()).foundationValue as? Map<String, Any?>
            ?: error("$name: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "$name version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun cases(name: String, key: String): List<Map<String, Any?>> {
        val list = root(name)[key] as? List<Map<String, Any?>> ?: error("$name: no $key[]")
        assertTrue(list.isNotEmpty(), "$name $key corpus must not be empty")
        return list
    }

    /** Map equality must not hinge on which integer type a number happened to parse or fold as -
     *  Int, Long and a whole Double all normalize to Long, on both sides of every comparison. */
    private fun normalized(value: Any?): Any? = when (value) {
        is Int -> value.toLong()
        is Double -> if (value == floor(value)) value.toLong() else value
        is Map<*, *> -> value.entries.associate { (key, entry) -> key as String to normalized(entry) }
        is List<*> -> value.map { normalized(it) }
        else -> value
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun wireConstantsAgreeWithCorpus() {
        val constants = root("wire.json")["constants"] as? Map<String, Any?> ?: error("wire.json: no constants")
        assertEquals((constants["wireVersion"] as Number).toInt(), LiveLogs.WIRE_VERSION, "wireVersion")
        assertEquals((constants["messageCap"] as Number).toInt(), LiveLogs.MESSAGE_CAP, "messageCap")
        assertEquals((constants["batchMaxRows"] as Number).toInt(), LiveLogs.BATCH_MAX_ROWS, "batchMaxRows")
        assertEquals((constants["queueCap"] as Number).toInt(), LiveLogs.QUEUE_CAP, "queueCap")
        assertEquals((constants["idleAckPause"] as Number).toInt(), LiveLogs.IDLE_ACK_PAUSE, "idleAckPause")
        assertEquals((constants["ringCap"] as Number).toInt(), LiveLogs.RING_CAP, "ringCap")
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun rowFoldsAgreeWithCorpus() {
        for (case in cases("wire.json", "rows")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val entry = case["entry"] as? Map<String, Any?> ?: error("$name: no entry")
            val at = (case["at"] as Number).toLong()
            val row = when (case["kind"]) {
                "log" -> LiveLogs.rowFromLog(
                    entry["scheme"] as String, entry["level"] as String, entry["message"] as String, at,
                )
                "error" -> LiveLogs.rowFromError(
                    entry["scheme"] as String, entry["code"] as String, entry["message"] as String?,
                    entry["recoverable"] as Boolean, entry["origin"] as String, at,
                )
                "kernel" -> LiveLogs.rowFromKernel(entry["line"] as String, at)
                else -> error("$name: unknown kind ${case["kind"]}")
            }
            assertEquals(normalized(case["expect"]), normalized(row), name)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun batchBodyAgreesWithCorpus() {
        for (case in cases("wire.json", "batch")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val body = LiveLogs.batchBody(
                case["sid"] as String,
                (case["n"] as Number).toLong(),
                case["rows"] as List<Map<String, Any?>>,
            )
            if (case["expect"] != null) assertEquals(normalized(case["expect"]), normalized(body), name)
            (case["canonical"] as? String)?.let {
                assertEquals(it, LiveLogs.canonical(body), "$name: canonical")
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun ackFoldAgreesWithCorpus() {
        for (case in cases("wire.json", "ack")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val steps = case["steps"] as? List<Map<String, Any?>> ?: error("$name: no steps")
            val expect = case["expect"] as? List<Map<String, Any?>> ?: error("$name: no expect")
            assertEquals(steps.size, expect.size, "$name: one expectation per step")

            var state = AckState.start()
            steps.forEachIndexed { index, step ->
                val where = "$name: step $index"
                val at = (step["at"] as Number).toLong()
                val ack = step["ack"] as? Map<String, Any?>
                state = when {
                    ack != null -> AckState.fold(
                        state,
                        ack["ok"] as Boolean,
                        (ack["viewers"] as Number).toInt(),
                        (ack["ttlMs"] as Number).toLong(),
                        at,
                    )
                    step["expire"] == true -> AckState.expire(state, at)
                    else -> error("$where names no operation")
                }
                for ((key, want) in expect[index]) when (key) {
                    "idle" -> assertEquals((want as Number).toInt(), state.idle, "$where: idle")
                    "paused" -> assertEquals(want as Boolean, state.paused, "$where: paused")
                    "stopped" -> assertEquals(want as Boolean, state.stopped, "$where: stopped")
                    "reason" -> assertEquals(want as String, state.reason, "$where: reason")
                    "deadline" -> assertEquals((want as Number).toLong(), state.deadline, "$where: deadline")
                    else -> error("$where: unknown expectation key $key")
                }
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun deviceQueueAgreesWithCorpus() {
        for (case in cases("wire.json", "queue")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val steps = case["steps"] as? List<Map<String, Any?>> ?: error("$name: no steps")
            val expect = case["expect"] as? List<Map<String, Any?>> ?: error("$name: no expect")
            assertEquals(steps.size, expect.size, "$name: one expectation per step")

            val queue = LiveQueue<Any?>((case["cap"] as Number).toInt())
            steps.forEachIndexed { index, step ->
                val want = expect[index]
                val where = "$name: step $index"
                when {
                    step.containsKey("push") -> {
                        val result = queue.push(step["push"])
                        assertEquals((want["size"] as Number).toInt(), result.size, "$where: size")
                        assertEquals((want["dropped"] as Number).toInt(), result.dropped, "$where: dropped")
                    }
                    step["batch"] != null -> assertEquals(
                        want["batch"] as List<Any?>,
                        queue.batch((step["batch"] as Number).toInt()),
                        "$where: batch",
                    )
                    step["ack"] != null -> assertEquals(
                        (want["size"] as Number).toInt(),
                        queue.ack((step["ack"] as Number).toInt()),
                        "$where: size after ack",
                    )
                    else -> error("$where names no operation")
                }
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun relayRingAgreesWithCorpus() {
        for (case in cases("wire.json", "ring")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val steps = case["steps"] as? List<Map<String, Any?>> ?: error("$name: no steps")
            val expect = case["expect"] as? List<Map<String, Any?>> ?: error("$name: no expect")
            assertEquals(steps.size, expect.size, "$name: one expectation per step")

            val ring = LiveRing<Any?>((case["cap"] as Number).toInt())
            steps.forEachIndexed { index, step ->
                val want = expect[index]
                val where = "$name: step $index"
                when {
                    step["appendBatch"] != null -> {
                        val append = step["appendBatch"] as Map<String, Any?>
                        val result = ring.appendBatch((append["n"] as Number).toLong(), append["rows"] as List<Any?>)
                        assertEquals(want["accepted"] as Boolean, result.accepted, "$where: accepted")
                        assertEquals((want["last"] as Number).toLong(), result.last, "$where: last")
                        assertEquals(result.last, ring.last, "$where: last is the ring's cursor")
                    }
                    step["read"] != null -> {
                        val read = step["read"] as Map<String, Any?>
                        val result = ring.read((read["after"] as Number).toLong(), (read["limit"] as Number).toInt())
                        assertEquals(
                            (want["seqs"] as List<Any?>).map { (it as Number).toLong() },
                            result.rows.map { it.seq },
                            "$where: seqs",
                        )
                        assertEquals(want["gap"] as Boolean, result.gap, "$where: gap")
                    }
                    else -> error("$where names no operation")
                }
            }
        }
    }

    @Test
    fun sha256AgreesWithTheStandardVectors() {
        for (case in cases("report.json", "sha256")) {
            val input = case["input"] as String
            assertEquals(case["expect"] as String, LiveLogs.sha256Hex(input), "sha256(${input.take(32)})")
        }
    }

    @Test
    fun canonicalBytesAgreeWithCorpus() {
        for (case in cases("report.json", "canonical")) {
            val name = case["name"] as? String ?: "<unnamed>"
            assertEquals(case["expect"] as String, LiveLogs.canonical(case["value"]), name)
        }
    }

    @Test
    fun whatHasNoCanonicalFormIsRefused() {
        for (case in cases("report.json", "canonicalRejects")) {
            val name = case["name"] as? String ?: "<unnamed>"
            assertFailsWith<IllegalArgumentException>(name) { LiveLogs.canonical(case["value"]) }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theReferenceReportSealsToThePins() {
        for (case in cases("report.json", "seal")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val body = case["body"] as? Map<String, Any?> ?: error("$name: no body")
            val seal = LiveLogs.reportSeal(body)
            assertEquals(case["hash"] as String, seal.hash, "$name: hash")
            assertEquals(case["text"] as String, seal.text, "$name: text")
            val roundTrip = LiveLogs.reportVerdict(seal.text)
            assertEquals("genuine", roundTrip.verdict, "$name: a fresh seal must verify")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun verifierVerdictsAgreeWithCorpus() {
        for (case in cases("report.json", "verdict")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = case["expect"] as? Map<String, Any?> ?: error("$name: no expect")
            val verdict = LiveLogs.reportVerdict(case["text"] as String)
            assertEquals(expect["verdict"] as String, verdict.verdict, "$name: verdict")
            assertEquals(expect["assertion"] as Boolean, verdict.assertion, "$name: assertion")
        }
    }
}
